src/coll/mod.rs

```diff
@@ -96,53 +96,60 @@ impl Collection {
-        spec.insert("cursor", Bson::Document(cursor));
-        if opts.allow_disk_use {
-            spec.insert("allowDiskUse", Bson::Boolean(opts.allow_disk_use));
-        }
+        let mut spec = doc! {
+            "aggregate" => (self.name()),
+            "pipeline" => pipeline_map
+        };
+
+        let mut read_preference = self.read_preference.clone();
+
+        match options {
+            Some(aggregate_options) => {
+                if let Some(ref read_preference_option) = aggregate_options.read_preference {
+                    read_preference = read_preference_option.clone();
+                }
+
+                spec = merge_options(spec, aggregate_options);
+            }
+            None => {
+                let cursor = doc! { "batchSize" => (::cursor::DEFAULT_BATCH_SIZE) };
+                spec.insert("cursor", cursor);
+            }
+        };
 
-        let read_pref = opts.read_preference.unwrap_or(self.read_preference.to_owned());
-        self.db.command_cursor(spec, CommandType::Aggregate, read_pref)
+        self.db.command_cursor(spec, CommandType::Aggregate, read_preference)
     }
 
     /// Gets the number of documents matching the filter.
     pub fn count(&self,
                  filter: Option<bson::Document>,
                  options: Option<CountOptions>)
                  -> Result<i64> {
-        let opts = options.unwrap_or_else(CountOptions::new);
-
         let mut spec = bson::Document::new();
         spec.insert("count", Bson::String(self.name()));
-        spec.insert("skip", Bson::I64(opts.skip as i64));
-        spec.insert("limit", Bson::I64(opts.limit));
-        if filter.is_some() {
-            spec.insert("query", Bson::Document(filter.unwrap()));
+
+        if let Some(filter_doc) = filter {
+            spec.insert("query", Bson::Document(filter_doc));
         }
 
-        // Favor specified hint document over string
-        if opts.hint_doc.is_some() {
-            spec.insert("hint", Bson::Document(opts.hint_doc.unwrap()));
-        } else if opts.hint.is_some() {
-            spec.insert("hint", Bson::String(opts.hint.unwrap()));
+        let mut read_preference = self.read_preference.clone();
+
+        if let Some(count_options) = options {
+            if let Some(ref read_preference_option) = count_options.read_preference {
+                read_preference = read_preference_option.clone();
+            }
+
+            spec = merge_options(spec, count_options);
         }
 
-        let read_pref = opts.read_preference.unwrap_or(self.read_preference.to_owned());
-        let result = try!(self.db.command(spec, CommandType::Count, Some(read_pref)));
+        let result = try!(self.db.command(spec, CommandType::Count, Some(read_preference)));
         match result.get("n") {
-            Some(&Bson::I32(ref n)) => Ok(*n as i64),
-            Some(&Bson::I64(ref n)) => Ok(*n),
+            Some(&Bson::I32(n)) => Ok(n as i64),
```

Comment:
Numeric types (such as `i32` and `i64`) implement `Copy`, so there isn't any need to capture them by reference.